### ISSUE TITLE
Python support does not compile

### ISSUE THREAD

**Issue by rodrigob** [halide/Halide#814]
In current master swig has troubles parsing Halide.h (see #691).
Even if one "disables chunks of Halide.h" (limiting which area of the API is exposed to Python), then more issues follow.

I am currently stuck with
ImportError: Halide_master/python_bindings/halide/_cHalide.cpython-34m.so: undefined symbol: halide_error_buffer_argument_is_null

Neither `nm ../build/lib/libInitialModules.a` nor ` nm ../build/lib/libHalide.so` seem to define this symbol (despite being defined in Halide.h). 

What I am missing ?
Your help getting Python support back would be most welcome.

**Comment by zvookin** [comment]
This is part of the runtime, which is included in generated filters. I
would say this should be declared in HalideRuntime.h, not Halide.h. I'll
try to take a look at the code in a little bit to see how it is making it
into Halide.h.

If you need runtime symbols, probably the easiest way to get them is to
link a generated filter into your executable. Probably more of a
workaraound than a fix, but I mention it in case it gets you moved forward
for now.

-Z-

**Comment by zvookin** [comment]
Hmmm... nevermind, it looks like you have a libInitialModules.a mentioned
there. Where does that come from? The routine in question should be coming
from src/runtime/posix_error.cpp .

-Z-

**Comment by zvookin** [comment]
Sorry, src/runtime/posix_error_handler.cpp is the file. libInitialModules.a
appears to hold the initmods for the CMake build, where initmods are blocks
of llvm bitcode for the runtime and of course do not provide the routines
inside the bit code as exports. I think libInitialModules.a should be
included in libHalide.so anyway though perhaps the CMake build generates a
different library structure than the Makefile. (Which would be unfortunate
if true.)

I do not see how halide_error_buffer_argument_is_null is ending up in
Halide.h. Is it possible SWIG is seeing HalideRuntime.h as well? That might
make sense to do so Python code can call runtime routines in generated
filters, but none of those routines will be in libHalide.so. (I am trying
to build a current master so I can check Halide.h now, but this machine is
way behind on llvm as well so it will take a while.)

-Z-

**Comment by aam** [comment]
Regarding how halide_error_buffer_argument_is_null defined in runtime/HalideRuntime.h is ending up in Halide.h - I think the reason is that runtime/HalideRuntime.h is listed as one of the HEADER_FILES in Makefile:
```
HEADER_FILES = \
...
runtime/HalideRuntime.h \
...

include/Halide.h: $(HEADERS) src/HalideFooter.h $(BIN_DIR)/build_halide_h
	mkdir -p include
	cd src; ../$(BIN_DIR)/build_halide_h $(HEADER_FILES) HalideFooter.h > ../include/Halide.h; cd ..
```

**Comment by zvookin** [comment]
Yes, it is explicitly added to Halide.h in order to get the definition of
struct buffer_t . (I believe HalideRuntime.h would be included otherwise
because some other headers include it as well.) This is somewhat
problematic as not all the routines in the file are provided by the Halide
library. It also ends up include runtime_internal.h in Halide.h, which puts
a very tricky and ugly chunk of code in our public facing interface.

In addition to buffer_t, the routines in device_interface are available in
libHalide.a . I'd like to do this more cleanly, but my recollection is we
decided this was the better alternative when HalideRuntime.h was added to
Halide.h . I'll let others weigh in.

Is it possible to black list specific routines from SWIG's consideration?
Almost everything in HalideRuntime.h would need to be so black listed I
believe.

-Z-

On Sun, May 31, 2015 at 8:30 PM, Alexander Aprelev <notifications@github.com

**Comment by rodrigob** [comment]
> Is it possible to black list specific routines from SWIG's consideration?
Yes, 80% of https://github.com/halide/Halide/blob/master/python_bindings/halide/cHalide.i is doing that.

I guess the real issue is that python compilation/execution is not in the build test suite; thus it has now happened that the progress in C++ part broke the python support.

**Comment by rodrigob** [comment]
I tried black listing HalideRuntime.h (adding `#if !defined(SWIG)`), but I saw no change in the result (to be double checked).

What would be strategy to do it the other way around. Is it there a "libHalideRuntime.a" that one could simply link to ?

**Comment by zvookin** [comment]
The easiest way to get a runtime is to generate a small filter into a .o
and link to it. We've talked about providing the entire runtime as a
library which could be used in the JIT case, but it requires answering both
design and build questions. Generally for JIT, the runtime is made on the
fly when realize is called. My biggest issue with having a standalone
library for the complete runtime is that it multiplies the testing load by
2x. There hasn't been a good use case for it either and with the modules
changes, we now have a pretty good model for sharing or not sharing a
single runtime as needs dictate.

It would be good to know where the reference is coming from if not
HalideRuntime.h . I assume SWIG is generating code to call this routine. Is
that the case? If so, it should be possible to see what is prompting SWIG
to do so. No code in libHalide.a would call this, only code generated by
calling realize or compile_to_file, etc.

-Z-

**Comment by rodrigob** [comment]
I understand.
Then I will double check my last test.
There is no need to track who includes HalideRuntime.h when I modify it by including `#if !defined(SWIG)`. With that if condition, the whole content of the file is ignored any time SWIG reaches it.

**Comment by rodrigob** [comment]
To get things going I have created #821

**Comment by rodrigob** [comment]
#821 is now feature complete and is a viable option to replace the swig interface.

**Comment by rodrigob** [comment]
Fixed with #821.
